src/combinatorics/simple_graph/subgraph.lean

```diff
@@ -0,0 +1,296 @@
+/-
+Copyright (c) 2021 Hunter Monroe. All rights reserved.
+Released under Apache 2.0 license as described in the file LICENSE.
+Authors: Hunter Monroe, Kyle Miller, Alena Gusakov
+-/
+import combinatorics.simple_graph.basic
+import data.set.finite
+
+/-!
+# Subgraphs of a simple graph
+
+A subgraph of a simple graph consists of subsets of the graph's vertices and edges such that the
+endpoints of each edge are present in the vertex subset.  The edge subset is formalized as a
+sub-relation of the adjacency relation of the simple graph.
+
+## Main definitions
+
+* `subgraph G` is the type of subgraphs of a `G : simple_graph`
+
+* `subgraph.neighbor_set`, `subgraph.incidence_set`, and `subgraph.degree` are like their
+  `simple_graph` counterparts, but they refer to vertices from `G` to avoid subtype coercions.
+
+* `subgraph.coe` is the coercion from a `G' : subgraph G` to a `simple_graph G'.verts`.
+  (This cannot be a `has_coe` instance due to universe variable inference issues.)
+
+* A `bounded_lattice (subgraph G)` instance.
+
+* Graph homomorphisms from a subgraph to a graph (`subgraph.map_top`) and between subgraphs
+  (`subgraph.map`).
+
+## Todo
+
+* Images of graph homomorphisms as subgraphs.
+
+-/
+
+universe u
+
+namespace simple_graph
+
+/-- A subgraph of a `simple_graph` is a subset of vertices along with a restriction of the adjacency
+relation that is symmetric and is supported by the vertex subset.  They also form a bounded lattice.
+
+Thinking of `V → V → Prop` as `set (V × V)`, a set of darts, then `subgraph.adj_sub` is that the
```

Comment:
Darts in graph theory are a name for the half-edges.  I'll add a comment.

Everyone wants `subgraph` to be a `set_like` :smile:   Subgraphs are not determined by their vertex sets  since the edges can vary.  I'll add implementation notes about this.